Assets/Code/data classes/visibilty.cs

```diff
@@ -3,112 +3,189 @@
+  {
+      planet = 1,
+      satellite = 2,
+      facility = 4,
+      all = 8
+  }
 
-  public bool raycast(position position1, position position2, int flags, int defaultRadius = 1)
+  public struct raycastInfo
+  {
+
+    public bool hit;
+    public List<planet> hitPlanets;
+    public List<facility> hitFacs;
+    public List<satellite> hitSats;
+
+    public raycastInfo(bool hit, List<planet> HitPlanets, List<satellite> HitSats, List<facility> HitFacs)
+    {
+      this.hit = hit;
+      this.hitPlanets = HitPlanets;
+      this.hitSats = HitSats;
+      this.hitFacs = HitFacs;
+    }
+  }
+
+  public static double distance(position p1, position p2)
+  {
+     double distance = Math.Sqrt(Math.Pow((p1.x - p2.x) ,2) + Math.Pow((p1.y - p2.y) ,2) + Math.Pow((p1.z - p2.z), 2));
+     return distance;
+  }
+
+  public static double raycastMath(position position1, position position2, position obj)
+  {
+    double distance1 = distance(position1, obj);
+    double distancePercent = distance1 / distance(position1, position2);
+
+    position pointOnLine = new position((position1.x + position2.x) * distancePercent,
+    (position1.y + position2.y) * distancePercent,
+    (position1.z + position2.z) * distancePercent);
+
+    double distanceFromObj = distance(pointOnLine, obj);
+
+    return distanceFromObj;
+
+  }
+
+  //double.max
+
+  public static raycastInfo raycastHit(position p1, position p2, raycastParameters flags, int defaultRadius)
   {
     // 100 indicates planet search
     // 10 indicates a satellite search
     // 1 indicates a facility search
 
-    int c = 0;
-    int x = flags;
-    bool check = false;
-
+    List<planet> hitPlanets = new List<planet>();
+    List<facility> hitFacs = new List<facility>();
+    List<satellite> hitSats = new List<satellite>();
+    bool hit = false;
 
-    if (x - 100 >= 0)
+    if ((flags & raycastParameters.planet) == raycastParameters.planet)
     {
-      x = x - 100;
-      foreach (planet p in master.allPlanets)
+      foreach(planet p in master.allPlanets)
       {
-        double distance1 = Math.Sqrt(Math.Pow((position1.x - p.pos.x) ,2) + Math.Pow((position1.y - p.pos.y) ,2) + Math.Pow((position1.z - p.pos.z), 2));
-        double distancePercent = distance1 / Math.Sqrt(Math.Pow((position1.x - position2.x), 2) + Math.Pow((position1.y - position2.y), 2) + Math.Pow((position1.z - position2.z), 2));
-
-        double pointOnLineX = (position1.x + ((position2.x - position1.x) * distancePercent));
-        double pointOnLineY = (position1.y + ((position2.y - position1.y) * distancePercent));
-        double pointOnLineZ = (position1.z + ((position2.z - position1.z) * distancePercent));
-
-        double inRadius = Math.Sqrt(Math.Pow((pointOnLineX - p.pos.x), 2) + Math.Pow((pointOnLineY - p.pos.y), 2) + Math.Pow((pointOnLineX - p.pos.z), 2));
-
-        if (inRadius < p.radius)
-        {
-          c = c + 1;
-        }
-        else
+        if (distance(p1, p.pos) > distance(p1,p2))
```

Comment:
You repeatedly use distance(p1, p2). Instead of recalc every if loop, save it in the beginning of the function then just reference it rather then recalc it.